cmd/testmatrix/main.go

```diff
@@ -0,0 +1,237 @@
+	scenarioPassRe = regexp.MustCompile(`: --- PASS:\s+([\w_]+)`)
+	scenarioSkipRe = regexp.MustCompile(`: --- SKIP:\s+([\w_]+)`)
+)
+
+func main() {
+	args := Args{}
+	arg.MustParse(&args)
+
+	ctx := context.Background()
+	cloudbuildService, err := cloudbuild.NewService(ctx)
+	if err != nil {
+		panic(err)
+	}
+	storageClient, err := storage.NewClient(ctx)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	listTriggersRes, err := cloudbuildService.Projects.Triggers.List(args.ProjectID).
+		Context(ctx).
+		PageSize(128).
+		Do()
+	if err != nil {
+		panic(err)
+	}
+
+	var wg sync.WaitGroup
+	results := make([]*result, len(listTriggersRes.Triggers))
+	for i, trigger := range listTriggersRes.Triggers {
+		wg.Add(1)
+		go handleTrigger(
+			ctx,
+			&wg,
+			args.ProjectID,
+			&results[i],
+			trigger,
+			cloudbuildService,
+			storageClient,
+		)
+	}
+	wg.Wait()
+
+	repoToPlatformToScenario := map[string]map[string]map[string]status{}
+	repoNameSet := map[string]struct{}{}
+	scenarioSet := map[string]struct{}{}
+	platformSet := map[string]struct{}{}
+	for _, result := range results {
+		if result == nil || result.Platform == "build" {
+			continue
+		}
+
+		repoNameSet[result.RepoName] = struct{}{}
+		if repoToPlatformToScenario[result.RepoName] == nil {
+			repoToPlatformToScenario[result.RepoName] = map[string]map[string]status{}
+		}
+		platformToScenario := repoToPlatformToScenario[result.RepoName]
+
+		platformToScenario[result.Platform] = result.Statuses
+		platformSet[result.Platform] = struct{}{}
+
+		for scenario := range result.Statuses {
+			scenarioSet[scenario] = struct{}{}
+		}
+	}
+	repoNames := sortStringSet(repoNameSet)
+	scenarios := sortStringSet(scenarioSet)
+	platforms := sortStringSet(platformSet)
+
+	template := template.Must(template.New("table").Parse(templateTxt))
+	err = template.Execute(os.Stdout, struct {
+		RepoNames                []string
+		Scenarios                []string
+		Platforms                []string
+		RepoToPlatformToScenario map[string]map[string]map[string]status
+		Pass                     status
+		Skip                     status
+	}{repoNames, scenarios, platforms, repoToPlatformToScenario, pass, skip})
+
+	if err != nil {
+		panic(err)
+	}
+}
+
+// handleTrigger populates the resPtr with the results of the latest build of
+// the given trigger.
+func handleTrigger(
+	ctx context.Context,
+	wg *sync.WaitGroup,
+	projectId string,
+	resPtr **result,
+	trigger *cloudbuild.BuildTrigger,
+	cloudbuildService *cloudbuild.Service,
+	storageClient *storage.Client,
+) {
+	if !triggerNameRe.MatchString(trigger.Name) {
+		return
+	}
+	res := result{
+		RepoName: trigger.Github.Name,
+		Platform: trigger.Tags[1],
+		Statuses: make(map[string]status),
+	}
+
+	// fetch the latest successful build
+	listRes, err := cloudbuildService.Projects.Builds.List(projectId).
+		Context(ctx).
+		Filter(fmt.Sprintf(`trigger_id="%v" AND status="SUCCESS"`, trigger.Id)).
+		PageSize(1).
+		Do()
+	if err != nil {
+		panic(err)
```

Comment:
Consider returning an error and using https://pkg.go.dev/golang.org/x/sync/errgroup in the caller.

Reply:
Looks much better, thanks